src/core/src/sketch/minhash.rs

```diff
@@ -516,67 +543,123 @@ impl KmerMinHash {
         Ok((it2.count() as u64, combined_mh.mins.len() as u64))
     }
 
-    pub fn compare(&self, other: &KmerMinHash) -> Result<f64, Error> {
-        self.check_compatible(other)?;
-        if let Ok((common, size)) = self.intersection_size(other) {
-            Ok(common as f64 / u64::max(1, size) as f64)
-        } else {
-            Ok(0.0)
+    // compare two minhashes, ignoring abundance.
+    pub fn compare(&self, other: &KmerMinHash, downsample: bool) -> Result<f64, Error> {
+        if downsample && self.max_hash != other.max_hash {
+            if self.max_hash < other.max_hash {
+                let mut new_mh = KmerMinHash::new(
+                    other.num,
+                    other.ksize,
+                    other.hash_function,
+                    other.seed,
+                    self.max_hash,
+                    other.abunds.is_some(),
+                );
+                new_mh.add_many(&other.mins)?;
+                return self.compare(&new_mh, false)
+            } else { // other.max_hash < self.max_hash
+                let mut new_mh = KmerMinHash::new(
+                    self.num,
+                    self.ksize,
+                    self.hash_function,
+                    self.seed,
+                    other.max_hash,
+                    self.abunds.is_some(),
+                );
+                new_mh.add_many(&self.mins)?;
+                return new_mh.compare(other, false)
+            }
         }
-    }
-
-    pub fn similarity(&self, other: &KmerMinHash, ignore_abundance: bool) -> Result<f64, Error> {
-        self.check_compatible(other)?;
-
-        if ignore_abundance {
+        else {
+            self.check_compatible(other)?;
             if let Ok((common, size)) = self.intersection_size(other) {
                 Ok(common as f64 / u64::max(1, size) as f64)
             } else {
                 Ok(0.0)
             }
-        } else {
-            if self.abunds.is_none() || other.abunds.is_none() {
-                // TODO: throw error, we need abundance for this
-                unimplemented!()
+        }
+    }
+
+    // compare two minhashes, with abundance.
+    pub fn similarity(&self, other: &KmerMinHash, ignore_abundance: bool, downsample: bool) -> Result<f64, Error> {
+        if downsample && self.max_hash != other.max_hash {
+            if self.max_hash < other.max_hash {
+                let mut new_mh = KmerMinHash::new(
+                    other.num,
+                    other.ksize,
+                    other.hash_function,
+                    other.seed,
+                    self.max_hash,
+                    other.abunds.is_some(),
+                );
+                //new_mh.add_many_with_abund(&other.abunds)?;
```

Comment:
`abunds` is a vector of `u64`, so we need to pass both `mins` and `abunds` here. I think there was `.to_vec()` method in `KmerMinhash` at some point, that would return pairs of `(hash, count)`, but if not I think it is a useful one to have and use in `add_many_with_abund`